libraries/AP_OSD/AP_OSD_ParamScreen.cpp

```diff
@@ -0,0 +1,663 @@
+            Vector3f vec = param->get();
+            vec.x = constrain_float(vec.x + incr, setting._param_min, setting._param_max);
+            vec.y = constrain_float(vec.y + incr, setting._param_min, setting._param_max);
+            vec.z = constrain_float(vec.z + incr, setting._param_min, setting._param_max);
+            param->set(vec);
+        }
+            break;
+        case AP_PARAM_NONE:
+        case AP_PARAM_GROUP:
+            break;
+        }
+    }
+}
+
+// modify which parameter is configued for the given selection
+void AP_OSD_ParamScreen::modify_configured_parameter(uint8_t number, Event ev)
+{
+    if (number > NUM_PARAMS) {
+        return;
+    }
+
+    _requires_save = true;
+
+    AP_OSD_ParamSetting& setting = params[number-1];
+    AP_Param* param;
+
+    if (ev == Event::MENU_DOWN) {
+        param = AP_Param::next_scalar(&setting._current_token, &setting._param_type);
+    } else {
+        // going backwards is somewhat convoluted as the param code is geared for going forward
+        ap_var_type type = AP_PARAM_NONE, prev_type = AP_PARAM_NONE, prev_prev_type = AP_PARAM_NONE;
+        AP_Param::ParamToken token, prev_token, prev_prev_token;
+
+        for (param = AP_Param::first(&token, &type);
+            param && (setting._current_token.key != token.key
+                || setting._current_token.idx != token.idx
+                || setting._current_token.group_element != token.group_element);
+            param = AP_Param::next_scalar(&token, &type)) {
+            prev_prev_token = prev_token;
+            prev_prev_type = prev_type;
+            prev_token = token;
+            prev_type = type;
+        }
+        if (param != nullptr) {
+            param = AP_Param::next_scalar(&prev_prev_token, &prev_prev_type);
+            setting._current_token = prev_prev_token;
+            setting._param_type = prev_prev_type;
+        }
+    }
+
+    if (param != nullptr) {
+        // update the stored index
+        setting._param_group = setting._current_token.group_element;
+        setting._param_key = AP_Param::get_persistent_key(setting._current_token.key);
+        setting._param_idx = setting._current_token.idx;
+        setting._param = param;
+        // force update() to refresh the token
+        setting._current_token.key = 0;
+        setting._current_token.idx = 0;
+        setting._current_token.group_element = 0;
+    }
+}
+
+// save all of the parameters
+void AP_OSD_ParamScreen::save_parameters()
+{
+    if (!_requires_save) {
+        return;
+    }
+
+    for (uint8_t i = 0; i < NUM_PARAMS; i++) {
+        if (params[i].enabled) {
+            AP_Param* p = params[i]._param;
+            if (p != nullptr) {
+                p->save();
+            }
+            params[i].save_as_new();
+        }
+    }
+    _requires_save = false;
+}
+
+// return radio values as LOW, MIDDLE, HIGH
+RC_Channel::AuxSwitchPos AP_OSD_ParamScreen::get_channel_pos(uint8_t rcmapchan) const
```

Comment:
We have an equivalent function on RC_Channel now.

Reply:
They are not quite the same and it didn't seem worth refactoring to pass in a bunch of different parameters for my use case.

Reply:
Please provide a really big comment here explaining that the ranges are different to the RC_Channel library, and why.